---
title: "Test MuSiC on pseudo-bulk data"
author: "Chiara Caprioli"
date: "`r Sys.Date()`"
output:
  html_document: 
   code: hide
---

MuSiC (https://doi.org/10.1038/s41467-018-08023-x) uses a scRNA reference to deconvolve 
bulk RNA-seq samples and obtain proportions of cell types in each sample.
The method is focused on two fundamental types of consistency: cross-subject and cross-cell.
The first is to guard against bias in subject selection, and the second is to guard 
against bias in cell capture in scRNA-seq. 
Rather than pre-selecting marker genes from scRNA-seq based only on mean expression, 
MuSiC gives weight to each gene, allowing for the use of a larger set of genes in deconvolution. 
The weighting scheme prioritizes consistent genes across subjects: 
up-weighing genes with low cross-subject variance (informative genes) 
and down-weighing genes with high cross-subject variance (non-informative genes). 
This requirement on cross-subject consistency is critical for transferring cell type-specific 
gene expression information from one dataset to another.
To deal with collinearity, MuSiC employs a tree-guided procedure that recursively zooms 
in on closely related cell types. First, similar cell types are grouped into the 
same cluster and cluster proportions are estimated; then, this procedure is recursively repeated 
within each cluster. At each recursion stage, only use genes that have low within-cluster variance, 
a.k. a. the cross-cell consistent genes, are used.

For benchmarking, we test MuSiC in 4 different modalities:
- direct with no marker selection
- direct with prior marker selection (cell type-specific marker genes by Wilcoxon)
- recursive with no marker selection
- recursive with prior marker selection (cell type-specific marker genes by Wilcoxon)

The above is performed on two different sets of pseudo-bulk data:
- summed counts
- TPM counts

```{r}
# Libraries
#devtools::install_github('xuranw/MuSiC', force = TRUE) # integrated package for both MuSiC and MuSiC2
suppressPackageStartupMessages({
library(MuSiC)
library(Seurat)
library(SingleCellExperiment)
library(Biobase)
library(ggpubr)
library(tidyverse)
library(ggdendro)
})

# Utils
wdir <- "~/Desktop/working/bulk_ct_deconvolution/"
outdir <- paste0(wdir, "benchmarking/MuSiC") 
if (!dir.exists(outdir)) {
  dir.create(outdir)
}
outdir2 <- paste0(wdir, "benchmarking/results")
source(paste0(wdir, "utils.R"))
```

# Data preparation
```{r}
## Bulk
### summed
pb_count_summed <- readRDS(paste0(wdir, "sc_reference/pb_data_sample_es_summed.rds"))
bulk.mtx.summed = exprs(pb_count_summed)

### TPM
pb_count_tpm <- readRDS(paste0(wdir, "sc_reference/pb_data_sample_es_tpm.rds"))
bulk.mtx.tpm = exprs(pb_count_tpm)

### List of bulk matrices
L_bulk <- list("summed" = bulk.mtx.summed, "tpm" = bulk.mtx.tpm)

## Single-cell
seurat <- readRDS(paste0(wdir, "sc_reference/sc_reference_sample.rds"))
sce <- as.SingleCellExperiment(seurat)

## Markers
markers <- read_csv(paste0(wdir, "sc_reference/sc_ct_markers.csv"), show_col_types = FALSE)
selected_markers <- markers %>%
  filter(p_val_adj < 0.05 & avg_log2FC > 2) 
```

# Bulk Tissue Cell Type Estimation
## Direct (not dealing with collinearity)
```{r}
direct_est_prop <- BenchmarkMusicDirect(
  L_bulk = L_bulk, 
  sc.sce = sce, 
  clusters = "aggregated_lineage2", 
  L_markers = L_markers,
  samples = "sample"
  )

saveRDS(direct_est_prop, paste0(outdir2, "/MuSiC.direct_proportions.rds"))
```

## Recursive (dealing with collinearity)
```{r}
L <- GetCellTypeHierarchies(
  x = sce, 
  clusters = "aggregated_lineage2", 
  markers = unique(selected_markers$gene), 
  samples = "sample", 
  option = "both"
  )

ggsave(paste0(outdir, "/hierarchies_no_markers.png"), L[[1]], width = 4, height = 4)
ggsave(paste0(outdir, "/hierarchies_with_markers.png"), L[[2]], width = 4, height = 4)
```

```{r}
# no markers
clusters.type_no_markers = list(
  C1 = c("Early_myeloid", "HSC_MPP"),
  C2 = c("DC", "Mono", "Late_myeloid"), 
  C3 = c("B_mature", "NK_cells", "T_CD4", "T_CD8"),
  C4 = "Erythroid", 
  C5 = "B_early",
  C6 = "Eo_baso_mast"
  )

cl.type = as.character(sce$aggregated_lineage2)

for(cl in 1:length(clusters.type_no_markers)){
  cl.type[cl.type %in% clusters.type_no_markers[[cl]]] = names(clusters.type_no_markers)[cl]
}
sce$clusterType_no_markers = factor(cl.type, levels = names(clusters.type_no_markers))
grouped <- unlist(clusters.type_no_markers)

seurat <- as.Seurat(sce)
Idents(seurat) <- seurat$clusterType_no_markers

group_markers1 <- FindAllMarkers(
  seurat,
  logfc.threshold = 0.5,
  test.use = "wilcox",
  slot = "data",
  min.pct = 0.3,
  only.pos = TRUE,
  min.cells.feature = 3,
  min.cells.group = 3
) 
group_markers1 <- group_markers1[which(group_markers1$p_val_adj < 0.05),]

group_markers1 = list(
  group_markers1$gene[which(group_markers1$cluster == "C1")],
  group_markers1$gene[which(group_markers1$cluster == "C2")],
  group_markers1$gene[which(group_markers1$cluster == "C3")],
  group_markers1$gene[which(group_markers1$cluster == "C4")],
  group_markers1$gene[which(group_markers1$cluster == "C5")],
  group_markers1$gene[which(group_markers1$cluster == "C6")]
)
names(group_markers1) = c('C1', 'C2', 'C3', 'C4', 'C5', 'C6')

# with markers
clusters.type_with_markers = list(
  C1 = c("Early_myeloid", "HSC_MPP"),
  C2 = c("DC", "Mono", "Late_myeloid"), 
  C3 = c("NK_cells", "T_CD4", "T_CD8"),
  C4 = c("B_mature", "B_early"), 
  C5 = "Erythroid",  
  C6 = "Eo_baso_mast"
)

cl.type = as.character(sce$aggregated_lineage2)

for(cl in 1:length(clusters.type_with_markers)){
  cl.type[cl.type %in% clusters.type_with_markers[[cl]]] = names(clusters.type_with_markers)[cl]
}
sce$clusterType_with_markers = factor(cl.type, levels = names(clusters.type_with_markers))
grouped <- unlist(clusters.type_with_markers)

seurat <- as.Seurat(sce)
Idents(seurat) <- seurat$clusterType_with_markers

group_markers2 <- FindAllMarkers(
  seurat,
  logfc.threshold = 0.5,
  test.use = "wilcox",
  slot = "data",
  min.pct = 0.3,
  only.pos = TRUE,
  min.cells.feature = 3,
  min.cells.group = 3
) 
group_markers2 <- group_markers2[which(group_markers2$p_val_adj < 0.05),]

group_markers2 = list(
  group_markers2$gene[which(group_markers2$cluster == "C1")],
  group_markers2$gene[which(group_markers2$cluster == "C2")],
  group_markers2$gene[which(group_markers2$cluster == "C3")],
  group_markers2$gene[which(group_markers2$cluster == "C4")],
  group_markers2$gene[which(group_markers2$cluster == "C5")],
  group_markers2$gene[which(group_markers2$cluster == "C6")]
)
names(group_markers2) = c('C1', 'C2', 'C3', 'C4', 'C5', 'C6')

L_markers <- list("no_markers" = group_markers1, "with_markers" = group_markers2)
L_clusters.type <- list("no_markers" = clusters.type_no_markers, "with_markers" = clusters.type_with_markers)
```

```{r}
recursive_est_prop <- BenchmarkMusicRecursive(
  L_bulk = L_bulk, 
  L_markers = L_markers, 
  L_clusters.type = L_clusters.type,
  sc.sce = sce, 
  clusters = "aggregated_lineage2", 
  samples = "sample"
  )

saveRDS(recursive_est_prop, paste0(outdir2, "/MuSiC.recursive_proportions.rds"))
```

## 5. Session info
```{r}
sessionInfo()
```

## 6. References
- Wang X et al., Bulk tissue cell type deconvolution with multi-subject single-cell expression reference. https://doi.org/10.1038/s41467-018-08023-x 
- https://xuranw.github.io/MuSiC/articles/MuSiC.html 
